include/rocksdb/file_system.h

```diff
@@ -825,6 +852,20 @@ class FSRandomAccessFile {
     return IOStatus::NotSupported("InvalidateCache not supported.");
   }
 
+  //  Asynchronous Read of the data requested in FSReadRequest. This is a
+  //  asynchronous call, i.e it will return after submitting the request.
+  //  When read request is completed callback function specified in cb will be
+  //  called with result populated in FSReadResponse. This API also populates
+  //  IOHandle. req contains the request offset and size, but the result and
+  //  status fields should be ignored. Instead, they should be passed in the
+  //  FSReadResponse structure to the callback.
+  virtual IOStatus ReadAsync(const IOOptions& /*opts*/, IODebugContext* /*dbg*/,
+                             FSReadRequest* /*req*/,
+                             std::function<void(FSReadResponse* resp)> /*cb*/,
+                             IOHandle* /*io_handle*/) {
```

Comment:
It's interesting that we are supporting both callback and polling. Is it very hard to implement? Thinking about the data race between the two, it feels not easy.

Reply:
I didn't think about this, but its an interesting idea. I guess polling and async callback initially made sense because PosixEnv requires polling the IO uring, whereas warm storage can complete an IO asynchronously. But a warm storage Poll() implementation could handle the required synchronization by itself and just return the results.